### PR TITLE
Feature branch - Problem2

### PR DESCRIPTION
**Pull request by intellinor**
Aici am complicat putin treaba intentionat, se putea mult mai simplu, dar voiam sa scot cat mai putine iteratii, cu acest minunat limbaj #fun :smiley:

### CODE REVIEW

**Review by alexandru-calinoiu — CHANGES_REQUESTED** [review]
You missed the point of TDD, test driven development in a nutshell means you have to run through the r-g-refactor loop many times in short iterations.

So for this problem for example a chain of commits will be:

- (red) Add test for empty array
- (green) Empty array returns no maximums
- (red) Add test for array with one element
- (green) Return both maximus as the only element in the array
- (refactor) Rename calculate method

File: src/problem2.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    compareValues maxValues,firstValue
+    if secondValue != nil
+      compareValues maxValues,secondValue
+    end
+  end
+
+  def compareValues(maxValues, newValue)
+
+    if maxValues[0] <= newValue || maxValues[1] <= newValue
+      replaceMaxNumbers maxValues, newValue
```
Comment (pirvudoru) [inline src/problem2.rb]: `compareValues` does much more than it advertises (would expect it to return a comparison result aka one of [-1, 0, 1]). In this case it mutates one of the parameters.

File: src/problem2.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+class Problem2
+  def calculateMaximum(values) #theta(n/2) -> O(n)
```
Comment (pirvudoru) [inline src/problem2.rb]: A good practice for comments (unless used for documentation) is to denote `why`, not `what`.

File: src/problem2.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  def setMaxNumbers(maxValues, firstValue, secondValue)
+    if maxValues[0] == nil && maxValues[1] == nil
+      maxValues[0] = firstValue
+      maxValues[1] = secondValue.nil? ? firstValue : secondValue
+      return
+    end
+    compareValues maxValues,firstValue
+    if secondValue != nil
+      compareValues maxValues,secondValue
```
Comment (pirvudoru) [inline src/problem2.rb]: Can you find the cause for calling `compareValues` twice and the `nil` checks ?

File: tests/problem2Tests.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  def test_equalMaxValueHasExpectedResult
+    assert_equal([1111,1111],@problem2.calculateMaximum(@testEqualMaxValueData))
+  end
+
+  def test_emptyArrayHasExpectedResult
+    assert_equal([nil,nil],@problem2.calculateMaximum(@testDataEmptyArray))
+  end
+
+  def test_oneElementArrayHasExpectedResult
+    assert_equal([9,9],@problem2.calculateMaximum(@testDataHasOneElement))
```
Comment (pirvudoru) [inline tests/problem2Tests.rb]: Don't believe that this is correct. You only have 1 number in the array.

**Review by pirvudoru — CHANGES_REQUESTED** [review]

File: tests/problem2Tests.rb (hunk 4 above)
Comment (intellinor) [reply tests/problem2Tests.rb]: In the previous implementation, I was returning [9,null].
Then I read this comment:

(red) Add test for array with one element
(green) Return both maximus as the only element in the array

Comment (pirvudoru) [reply tests/problem2Tests.rb]: :+1: Talked with @alexandru-calinoiu and any solution is ok since the problem statement can be interpreted in multiple ways. No change required here.

File: src/problem2.rb (hunk 3 above)
Comment (intellinor) [reply src/problem2.rb]: It's because i take numbers two by two, and the null check is required when we encounter arrays of odd length. This whole solution could have been much simpler if i would have taken numbers one by one.

File: src/problem2.rb (hunk 1 above)
Comment (intellinor) [reply src/problem2.rb]: evaluateValues or a similar name would be more clear maybe.

**Review by alexandru-calinoiu — CHANGES_REQUESTED** [review]
Is getting better,

But is still very far of, you are using a lot of procedural code with impure functions.

If you like to go down this rode and break this problem into smaller functions than readup https://en.wikipedia.org/wiki/Pure_function and make all your functions pure.
